google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/dialect/ForeignKeyTest.java

```diff
@@ -0,0 +1,183 @@
+        item.setItemId(1L);
+        item.setCart(cart);
+
+        CartSession cartSession = new CartSession();
+        cartSession.setSessionId(1L);
+        cartSession.setCart(cart);
+
+        cart.addItem(item);
+        cart.setSession(cartSession);
+
+        session.persist(cart);
+        session.persist(item);
+        session.persist(cartSession);
+      });
+
+    verifyOnDeleteCascadeReferentialConstraints();
+  }
+
+  @Test
+  public void testForeignKeyConstraint_verifyValidInserts() {
+
+    doInHibernate(this::sessionFactory, session -> {
+      Cart cart = new Cart();
+      cart.setCartId(2L);
+      cart.setItems(new HashSet<>());
+
+      Item item = new Item();
+      item.setItemId(2L);
+      item.setCart(cart);
+
+      CartSession cartSession = new CartSession();
+      cartSession.setSessionId(2L);
+      cartSession.setCart(cart);
+
+      cart.addItem(item);
+      cart.setSession(cartSession);
+
+      session.persist(cart);
+      session.persist(item);
+      session.persist(cartSession);
+    });
+
+    // validate that records exist
+    doInHibernate( this::sessionFactory, session -> {
+      Item item = session.get( Item.class, 2L );
+      assertEquals(2L, item.getCart().cartId);
+      CartSession cartSession = session.get( CartSession.class, 2L );
+      assertEquals(2L, cartSession.getCart().cartId);
+    } );
+  }
+
+  @Test
+  public void testForeignKeyConstraint_verifyInvalidInserts() {
+
+    final PersistenceException ex = assertThrows(
+        PersistenceException.class, () -> doInHibernate(this::sessionFactory, session -> {
+          Cart cart = new Cart();
+          cart.setCartId(3L);
+          cart.setItems(new HashSet<>());
+
+          Item item = new Item();
+          item.setItemId(3L);
+          item.setCart(cart);
+
+          CartSession cartSession = new CartSession();
+          cartSession.setSessionId(3L);
+          cartSession.setCart(cart);
+
+          cart.addItem(item);
+          cart.setSession(cartSession);
+
+          session.persist(item);
+        }));
+    assertThat(ex.getCause().getCause().getMessage()).contains("Foreign key constraint `Fk_itemDetails_cartId` is "
+        + "violated on table `Item`");
+  }
+
+  @Test
+  public void testForeignKeyConstraint_verifyDeletingReferencedKey() throws Exception {
+
+    doInHibernate(this::sessionFactory, session -> {
+      Cart cart = new Cart();
+      cart.setCartId(4L);
+      cart.setItems(new HashSet<>());
+
+      Item item = new Item();
+      item.setItemId(4L);
+      item.setCart(cart);
+
+      CartSession cartSession = new CartSession();
+      cartSession.setSessionId(4L);
+      cartSession.setCart(cart);
+
+      cart.addItem(item);
+      cart.setSession(cartSession);
+
+      session.persist(cart);
+      session.persist(item);
+      session.persist(cartSession);
+
+      // Delete the referenced cart object referenced by Item/CartSession child tables.
+      session.delete(cart);
+
+    });
```

Comment:
Sure breaking into two transactions failed since deleting and reading with the same transaction did not yield null value in the query. I broke it into 3 transactions. Deleting in the 2nd and querying back in the 3rd transaction. That gave the expected behaviour.